### DISCUSSION TITLE
Community Resources

### DISCUSSION THREAD

**Discussion by LexiconCode** [dictation-toolbox/Caster#237] [Ideas]
This is a place to post things that you find useful with the intent of adding them into a section of documentation of Caster for resources. I'm not talking about supported applications in Caster although if you have add-ons do mention them. Everything from guides, websites, repositories, software, hardware, and workflows that enhance your productivity or knowledge of programming. 

When posting a resource include:

- A link to the resource.
- Whether it's open source, commercial, personal use if applicable.
- Cost of ownership cost if applicable.
- How long have you been using the use resource and frequency that you use it.
- Optionally are you currently seeking for alternatives the posted resource? Have you used any other alternatives before finding the resource?
- Most importantly your experience with or what you get out of the resource personally? Try not the state a feature list but the features you find useful and expound on that. Any pros and cons?
- Try to maintain similar formatting as in the examples below.
  If you update your resource please mention `LexiconCode`. The state the date and content that you edited. 

**Community Resources**

**Apps**

By [@LexiconCode](https://github.com/LexiconCode) 

- [Fast Keys](https://www.fastkeysautomation.com/) Commercial, 19$ for 3 computers lifetime license. 
  I've used Fast Keys as a text expander for 4 years daily. I bought it for $10. Most of my dictation is done by voice but for the bit that I do interact with the computer is augmented by Fast Keys. I have created a shorthand for filling in emails and forms. Other aspects I appreciate is the simple interface, light on resources (RAM 2 MB), extensibility, not being integrated into the cloud. I've before purchasing Fast Keys I looked into/experimented with Text Expander, Phrase Expander, Breevy. I didn't know about at the time of purchase of Fast Keys Autohotkey as a open source alternative. 

- [GitExtensions ](https://gitextensions.github.io/) Open Source Free
  Almost daily for about a year and it has boosted my productivity. Unofficially Supported with DNS Select-and-Say for enhanced navigation menus and text. GitExtensions a GUI interface as a replacement for commandline for Git. While many IDE's/editors now integrate Git commands I still find myself coming back to GitExtensions. The program is well laid out and intuitive. I spend less time thinking about the commands of git and more about the functionality that I need. I wish it had more default keyboard commands. What I most appreciate most actions can be easily reverted. I'm overall inexperienced with Git and often become frustrated. This is a salve for weary bones.

- [Window Focus Logger](http://www.adminscope.com/downloads/window-focus-logger/) - Close Source - Free for personal and commercial use
	I use this to help debug application stealing focus unexpectedly. It can run in the background logging windows including window titles and process's location. Perhaps the most useful feature is it that only tells you in a interface is in focus but if it's keyboard focused as well.

- [Hunt and Peck](https://github.com/zsims/hunt-and-peck) Windows/[vimac](https://github.com/dexterleng/vimac) Mac - open source 
Simple vimium/vimperator style navigation for applications. 

![Screenshot](https://camo.githubusercontent.com/89b2faed2c93fbf77d32de171f305c47131127874b94f0428cfb94557a357988/68747470733a2f2f7261772e6769746875622e636f6d2f7a73696d732f68756e742d6e2d7065636b2f6d61737465722f73637265656e73686f74732f6578706c6f7265722e706e67)

By @LazoCoder 

- [Vimium](https://chrome.google.com/webstore/detail/vimium/dbepggeogbaibhgnhhndojpepiihcmeb?hl=en)(Plug-in for Google Chrome) Open Source Free

  This is a plug-in for Google Chrome that I use everyday. It finds all the links on a webpage and highlights them so that you can select them with letters on the keyboard (or voice) instead of clicking them with a mouse. It makes browsing the web by voice quick and seamless. I use it when I'm making Google searches, browsing Reddit, email, and pretty much everything else on the internet. Prior to discovering this plug-in, I used to use the "douglas" and "legion" grid systems to click things in the browser, but this is incredibly slow in comparison to Vimium.

The setup that I personally use is as follows. Go into Vimium's settings and copy and paste the following into "Custom key mappings":

```
unmapAll
map \ LinkHints.activateModeToOpenInNewTab
map / LinkHints.activateMode
map s scrollPageUp
map d scrollPageDown
map ? showHelp
```

  Now you can say "slash" to open a link or "backslash" to open it in a new tab. Furthermore, uttering the words "sierra" and "delta" move the page up and down respectively. These four commands are the only ones that I use, and they make a world of difference for me.

- Surfingkeys a Plug-in for [FireFox](https://addons.mozilla.org/en-US/firefox/addon/surfingkeys_ff/ )/[Google Chrome](https://chrome.google.com/webstore/detail/surfingkeys/gfbliohnnapiefjpjlpjnehglfpaknnc?hl=en-US) - Open Source Free

  People may already be aware of it, but I've got a fair bit of use out of an extension called SurfingKeys, which provides vim-like shortcuts for a lot of browser functionality, including hints for links and input fields, searching through bookmarks, moving the current tab to a new window, and a load of other stuff.

By @TTeig 

+ [+7 Taskbar Number](https://rammichael.com/7-taskbar-numberer) - Close source - Free

![](https://rammichael.com/wp-content/uploads/2012/07/7-Taskbar-Numberer.png)

  I have used it every day for about a month. It numbers your windows and tray icons and makes it so much easier to navigate between them. I was hesitant because there was no source code or license, but it was just too useful. I would like something more customizable and open source, but it does the job and I haven't seen any alternatives. (There are also other little useful tools from the creator that I haven't tried)

 **Workflow**

By @LexiconCode 

- To rebase **all** local branches on upstream/master in git bash: - [Source](https://wtanaka.com/node/8244)

```
BASEBRANCH=upstream/master
git for-each-ref --shell --format="git rebase $BASEBRANCH %(refname:short)" refs/heads/ | sh -x
```

Often I have a lot of working branches the need to be up-to-date with the development branch of Caster. This cuts out some of the time of rebasing but currently does not push to the origin branches. So enhancements could be made. When experimenting **backup your local repository**. This may not work as intended if you have a local branch like `gh-pages` with `C:\NatLink\NatLink\MacroSystem` your working directory git.

By @Versatilus 

- Procedures for maintaining Dragon NaturallySpeaking Accuracy

  One thing I do to help with accuracy whenever I make a new profile is that I dictate specific documents back into a text editor, correcting mistakes as I go. The [documents](https://www.google.com/search?q=speech+therapy+reading+passages+for+adults) are several short passages created by speech therapists specifically to help people with pronunciation. I'll look in a little while for links to a few of them.

  Additionally, I make a list of words that I consistently have issues with as I go and explicitly add them to the vocabulary (by saying "edit vocabulary"). I also make up random sentences containing the conflicting words and dictate those into a text editor. If I'm feeling proactive, I do this with other commands, even if I'm not specifically having problems with them.

  During this time I also spent extra time dictating things like documentation and detailed forum posts. I do most of my dictation in a text editor and make sure to save the files so that Dragon can use them to adapt to my speech patterns.

  It's important to frequently run Dragon's accuracy tuner during this process, then occasionally after that as well.

------

How to get help outside of an github issue.

- For those that prefer text-only utilize for general questions and to share your experience go to the [Caster Gitter Chat Channel](https://github.com/dictation-toolbox/Caster).

**Comment by LazoVelko** [comment]
[Vimium](https://vimium.github.io/) Open Source Free

This is a plug-in for Google Chrome that I use everyday. It finds all the links on a webpage and highlights them so that you can select them with letters on the keyboard (or voice) instead of clicking them with a mouse. It makes browsing the web by voice quick and seamless. I use it when I'm making Google searches, browsing Reddit, email, and pretty much everything else on the internet. Prior to discovering this plug-in, I used to use the "douglas" and "legion" grid systems to click things in the browser, but this is incredibly slow in comparison to Vimium.

The setup that I personally use is as follows. Go into Vimium's settings and copy and paste the following into "Custom key mappings":
```
unmapAll
map \ LinkHints.activateModeToOpenInNewTab
map / LinkHints.activateMode
map s scrollPageUp
map d scrollPageDown
map ? showHelp
```
Now you can say "slash" to open a link or "backslash" to open it in a new tab. Furthermore, uttering the words "sierra" and "delta" move the page up and down respectively. These four commands are the only ones that I use, and they make a world of difference for me.

**Comment by LexiconCode** [comment]
**Change Log**
 6-21- 2018 -  _Added_ - Apps: Vimium - **LazoCoder**  
10-16-2018 - _Added_  - Workflow: Maintaining Dragon NaturallySpeaking Accuracy - **Versatilus**
10-23-2018 - _Added_ - Workflow: Rebase all local branches - **LexiconCode** 
12-7-2019 - _Added_ - Surfingkeys Browser extensions - **mrob95** 
12-14-2019 - _Added_ - Window Focus Logger - **LexiconCode**
4-8-2020 - Added - +7 Taskbar Number - **TTeig**
12-28-2020 - _Added_ - Hunt and Peck - **LexiconCode**

**Comment by Digital-ecologist** [comment]
[kdocker](https://github.com/user-none/KDocker) Open source, no cost, compatible with Linux.
I've been using this utility for several days now. Caster with the Kaldi engine runs in a terminal. The terminal needs to be accessible for transparency and troubleshooting purposes. However, it can sometimes get in the way. I use kdocker to minimize the caster terminal to the system tray so that its unobtrusive but always accessible.
![system tray image](https://user-images.githubusercontent.com/42946430/109432832-1d38ea00-79db-11eb-976b-1de93c2589b0.png)

A word of caution: Kdocker will let you send anything to the system tray. I once accidentally sent my entire task bar to the system tray, and had to restart the system to get it back, because the system tray itself was hidden.

**Comment by hakanErgin** [comment]
![image](https://user-images.githubusercontent.com/50581015/116550812-7c41af00-a8f7-11eb-8763-5d5b55f2470d.png)
using [taskbar numberer](https://rammichael.com/7-taskbar-numberer) utility that was already mentioned above, I was able to generate numbers that continue after 9.

[_taskbar.zip](https://github.com/dictation-toolbox/Caster/files/6398545/_taskbar.zip)
and by using this script I was able to target the windows by their numbers

The script also supports commands for closing or minimizing and such, and is able to do similar actions for tray icons too

also by using [taskbar tweaker](https://rammichael.com/7-taskbar-tweaker) utility, I was able to manipulate the taskbar the way I want which I found very useful

**Comment by Digital-ecologist** [comment]
### Improved Windows commandline workflow combining Conemu & Cygwin
Tool 1: ConEmu [https://conemu.github.io/](url)
It's a free and open source terminal emulator for Windows. I've been using it for years to make the Windows command prompt more accessible. 
**Major advantages**
1. Normal integration with cut, copy, & paste functions
2. You can make the terminal output much easier to see with a custom font size and theme.  (I personally use Gamma 1 with 16 point font)
3. Integration with other terminal apps

Tool 2:  Cygwin [https://www.cygwin.com/](url)
It's also a free and open source tool. The primary function is to allow Bash and other Linux tools on Windows. In addition to the default tool set, it also lets you install additional packages such as fish. I used it for a bioinformatics project, then it sat on my system.

I recently discovered that you can set Cygwin Bash as your default shell in Conemu which allows interaction with the Windows system through Bash without any of the WSL set up.
![image](https://user-images.githubusercontent.com/42946430/116826786-bc0cce80-ab63-11eb-991e-7e4bc30995ff.png)

 Furthermore, you can call Fish  from within  Cygwin Bash within ConEmu. 
This gives access to really nice color coding, extremely powerful tab completion, and command history.
The end result is that you can use sauce, dunse, ross & tabby commands to save a ton of time when navigating your window system from the command line.
![image](https://user-images.githubusercontent.com/42946430/116829865-668ced80-ab74-11eb-9067-20843a35dc6d.png)
Hope this is helpful.

**Comment by Digital-ecologist** [comment]
Resource: Dragon Capture
Software type: Closed Source
Cost: $35
Main website: [https://www.rwilke.de/dragoncapture/]( https://www.rwilke.de/dragoncapture/)
Also available through Know-Brainer: [https://shop.knowbrainer.com/store/pc/viewCategories.asp?idCategory=103](https://shop.knowbrainer.com/store/pc/viewCategories.asp?idCategory=103)

I use this tool daily 8 hours a day on my work PC (because I'm constantly in applications not supported by Dragon). It also runs in the background on my personal laptop now, but I'm not sure how often it gets used, as it only steps in when necessary, and I'm not in unsupported applications as much.

[Edit] Thunderbird is the major application where this comes in handy for me on my personal laptop, and if you're a fan of of Liber office, (which I have on the system but don't use very much) it makes that feel more, although not really, native.

From the description on KB, it comes across as a glorified dictation box, but I find the real power lies in using it in auto paste mode for non-supported applications. The disadvantage there is that you don't get full Select-and-Say support, but since there non-supported applications to begin with,  you're not actually losing anything major in terms of Select-and-Say support, and the auto paste functionality is lightning-fast, relative to constantly summoning the standard dictation box.

R. Wilke who developed the tool has been super responsive whenever I've had questions on the KB forum.

One thing I have noticed after using it for a while, is that switching between the automatic and the dictation box replacement modes is a little bit cumbersome.

To address that, I've added the following as global commands in Dragon, to quickly toggle between the 2 modes. I suspect there is a more elegant way to do this than using heardword, but I haven't found it yet. There's a way of doing it with hot keys, but that has reliability issues depending on what application you have in focus, and how focus hungry that application is.

Spoken command:Box Mode
```
#Language "WWB-COM"

Option Explicit

Sub Main
HeardWord("show","box")
HeardWord("AutoPaste")
End Sub
```

Spoken command:Auto Mode

```
'#Language "WWB-COM"

Option Explicit

Sub Main
HeardWord("AutoPaste")
HeardWord("Allow","AutoPaste")
HeardWord("hide","box")
End Sub
```
If you spend most of your time and Dragon friendly applications, Dragon capture probably isn't worth it for you, however, if you find yourself having to spend a lot of time in applications that are not friendly to Dragon, or you find yourself using casters "say" & "cop" commands frequently to force dictation, you might find it really useful.
